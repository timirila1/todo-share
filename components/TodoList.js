import React, { Component } from 'react';
import { View, StyleSheet, ScrollView } from 'react-native';

import { TodoItem } from './TodoItem';
import { FloatButton } from './FloatButton';
import { TodoForm } from './TodoForm';
import { addTodo } from '../services/Todos';

export class TodoList extends Component {
    state = {
        addMode: false,
        todos: [],
        firstUpdate: true,
    }

    static getDerivedStateFromProps(props, state) {
        const { firstUpdate } = state;
        if (firstUpdate && props.todos.length > state.todos.length) {
            const { todos } = props;
            return { todos, firstUpdate: false };
        }
        const { todos } = state;
        return { todos };
    }

    getTodoItems = todos => todos.map(t => (
        <TodoItem
            todo={t}
            key={t.id}
        />
    ));


    setAddMode = value => this.setState({ addMode: value });

    addNewTodo = async (data) => {
        const { boardId } = this.props;
        const { todos } = this.state;
        const todo = { board_id: boardId, ...data };

        try {
            const newTodo = await addTodo(todo);
            this.setState({ todos: [...todos, newTodo] });
        } catch (e) {
            // todo add error handler
            console.error(e);
        }
        this.setAddMode(false);
    }

    render() {
        const { addMode, todos } = this.state;

        return (
            <View style={styles.container}>
                <ScrollView style={styles.container}>
                    {this.getTodoItems(todos)}
                    {addMode && <TodoForm
                        onCancel={() => this.setAddMode(false)}
                        onSend={data => this.addNewTodo(data)}
                    />}
                </ScrollView>
                {!addMode && <FloatButton
                    name="plus"
                    title="Add Todo"
                    onPress={() => this.setAddMode(true)}
                />}
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        paddingTop: 15,
    },
});
