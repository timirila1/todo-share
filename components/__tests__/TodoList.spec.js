import 'react-native';
import React from 'react';
import { TodoList } from '../TodoList';
import renderer from 'react-test-renderer';


it('Render TodoList', async () => {
    const todosMock = [{
        id: 1,
        title: 'Test title',
        desc: 'Test desc'
    },
    {
        id: 2,
        title: 'Test title',
        desc: 'Test desc'
    }];

    const tree = renderer.create(
        <TodoList
            navigation={{}}
            todos={todosMock}
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
