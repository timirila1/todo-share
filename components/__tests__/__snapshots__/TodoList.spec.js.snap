// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render TodoList 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "paddingLeft": 40,
        "paddingRight": 20,
        "paddingTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomColor": "#e9e9e9",
          "borderBottomWidth": 1,
          "flex": 1,
          "paddingBottom": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "300",
            "lineHeight": 36,
          }
        }
      >
        Test title
      </Text>
      <Text
        style={
          Object {
            "color": "#cacaca",
            "fontWeight": "200",
          }
        }
      >
        Test desc
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "paddingLeft": 40,
        "paddingRight": 20,
        "paddingTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomColor": "#e9e9e9",
          "borderBottomWidth": 1,
          "flex": 1,
          "paddingBottom": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "300",
            "lineHeight": 36,
          }
        }
      >
        Test title
      </Text>
      <Text
        style={
          Object {
            "color": "#cacaca",
            "fontWeight": "200",
          }
        }
      >
        Test desc
      </Text>
    </View>
  </View>
</View>
`;
